firedrake/functionspaceimpl.py

```diff
@@ -376,12 +376,18 @@ def make_function_space(cls, mesh, element, name=None):
         if mesh is not topology:
             # Create a concrete WithGeometry or FiredrakeDualSpace on this mesh
             new = cls.create(new, mesh)
+
+        if boundary_set:
+            new = RestrictedFunctionSpace(new, boundary_set=boundary_set)
```

Comment:
`RestrictedFunctionSpace` here is topological, so this should be done around line 374, I think.
At some point, I think we need to tidy up `reconstruct` and `make_function_space`.

Reply:
The way things are now, the geometric resticted space is wrapping a geometric unrestricted space:

```
from firedrake import *
mesh = UnitSquareMesh(1, 1)
V = FunctionSpace(mesh, "CG", 1)
problem = LinearVariationalProblem(inner(TrialFunction(V), TestFunction(V))*dx, 0, Function(V), bcs=DirichletBC(V, 0, [1]), restrict=True)

problem.u_restrict.function_space()

WithGeometry(RestrictedFunctionSpace("WithGeometry(FunctionSpace(<firedrake.mesh.MeshTopology object at 0x7ee09b678f0>, FiniteElement('Lagrange', triangle, 1), name=None), <Mesh #0>)", name=None, boundary_set=frozenset({1})), Mesh(VectorElement(FiniteElement('Lagrange', triangle, 1), dim=2), 0))
```